master/pkg/model/experiment_config_test.go

```diff
@@ -116,6 +116,22 @@ func TestOverrideMasterConfigRegistryAuth(t *testing.T) {
 	assert.DeepEqual(t, actual, expected)
 }
 
+func TestOverrideMasterEnvironmentVariables(t *testing.T) {
+	masterDefault := &TaskContainerDefaultsConfig{
+		EnvironmentVariables: &RuntimeItems{
+			CPU: []string{"a=from_master", "b=from_master"},
+		},
+	}
+	actual := DefaultEnvConfig(masterDefault)
+	assert.NilError(t, json.Unmarshal([]byte(`{
+    "environment_variables": ["a=from_exp", "c=from_master"]
+}`), &actual))
+	assert.DeepEqual(t, actual.EnvironmentVariables.CPU, []string{
+		"a=from_master", "b=from_master",
+		"a=from_exp", "c=from_master", // Exp config overwriters master config by being later.
```

Comment:
can you remind me, where does this happen? not in our code, but at runtime by the shell of the task container itself?

Reply:
Not sure if I fully understand the question but

In k8s it is our code converting the env list to a map taking precedence to later elements

In docker we pass the full list unprocessed into the docker runtime 

https://github.com/opencontainers/runtime-spec/blob/main/config.md#process

which states the env array has the same semantics as this document

https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap08.html#tag_08_01

Which technically says this behavior is undefined "There is no meaning associated with the order of strings in the environment. If more than one string in an environment of a process has the same name, the consequences are undefined."

Tried digging through Moby code to see if I could find where this behavior is defined but couldn't find it but I imagine somewhere it has this latest overrides earlier behavior

I don't know what do you think should we update the code to only pass in the latest variables to Docker?

Reply:
no worries, it's okay, I was just trying to remember if we do anything here or it's all external systems.